### ISSUE TITLE
Recover soft-deleted items

### ISSUE THREAD

**Issue by SukantGujar** [JKHeadley/rest-hapi#133]
**Is your feature request related to a problem? Please describe.**
Aside from serving as an archival process, soft-deletion also helps perform a recovery of the deleted data. Would an update query work on soft-deleted items to remove the deleted flag? Or would it require a MongoDB update?

**Describe the solution you'd like**
The update method can accept a flag to include soft-deleted items. This should allow their recovery.

**Comment by JKHeadley** [comment]
@SukantGujar currently this require's a MongoDB update. 

I agree this feature is needed, however I'm not sure if it should be included within the update endpoint since that would also allow documents to be 'deleted' through update.

I feel the best way would be to add a `recover` payload option to the `DELETE` endpoints. Similar to how the  [`hardDelete`](https://resthapi.com/docs/soft-delete.html) option works now.

**Comment by SukantGujar** [comment]
It might be a bit counter intuitive to use DELETE to recover a doc, but maybe it will make sense in the long run.
How about an explicit `POST /recover` endpoint in each model when soft-delete is enabled?

**Comment by JKHeadley** [comment]
@SukantGujar hmm, so you mean something like:
`POST /user/{_id}/recover` ?

**Comment by SukantGujar** [comment]
Yes
 
 
  On Fri, 29 Jun 2018 at 10:29, Justin Headley<notifications@github.com> wrote:   
@SukantGujar hmm, so you mean something like:
POST /user/{_id}/recover ?

**Comment by SukantGujar** [comment]
I am trying to implement this. This is the first time I am working on anything Hapi, so going really slow :). Couple of things:

1. Introduced two new config options: `enableRecoveryOfSoftDeletedItems` and `softDeletedRecoveryEndpoint`. The first one controls whether recovery endpoints are available and the second option controls the name of this endpoint, defaults to `recover` (POST /{modelname}/{_id}/recover). Sounds ok? This is different from what you suggested above.
2. While going through `rest-helper-factory` which contains the various API setup, noticed that the Auth scopes are also setup there. In this case, although the verb is `POST`, it makes sense to consider the same scope as `DELETE` for recovery also, considering a person who deleted would need to recover. This is simpler than introducing a new scope for recovery.
3. Will be adding a new method to the same file `generateRecoverOneEndpoint` and any supportive methods based on the `generateDeleteOneEndpoint` implementation.
4. Noticed that there's model validation code even for methods working on just the `id` of the resource (find and deleteone). Is this necessary?
5. Any other suggestions?

**Comment by prudradeep** [comment]
is this solved?

**Comment by prudradeep** [comment]
is this feature implemented?

**Comment by JKHeadley** [comment]
@prudradeep unfortunately it hasn't been implemented yet. 

@SukantGujar apologies for never getting back to you. Did you ever manage to make any progress here?

**Comment by prudradeep** [comment]
@JKHeadley  Can i help you on this

**Comment by JKHeadley** [comment]
@prudradeep absolutely! This feature will required a decent amount of code and testing but should look similar to the existing code so any help is greatly appreciated.

I agree with most of the points @SukantGujar made in the post above. Basically we will need two additional endpoints:

- `POST /{modelname}/recover/{_id}` to recover a single document.
- `POST /{modelname}/recover/` with a payload of an array of `_id`s to recover multiple documents.

This should look fairly similar to the `deleteOne` and `deleteMany` functions: https://github.com/JKHeadley/rest-hapi/blob/77eca88e0973d335f5183440ecc8ee6cf5257a43/utilities/handler-helper.js#L782

Please let me know if you have any other questions.

**Comment by prudradeep** [comment]
@JKHeadley This feature implemented on my local machine

Endpoints generated as follows:
- `PUT /{modelname}/{_id}/recover` to recover a single document
- `PUT /{modelname}/recover` with a payload of an array of `_id`s to recover multiple documents.

Above endpoints generated only if `enableSoftDelete` is enabled and `model.routeOptions.allowUpdate` is `true`.

As you mentioned these endpoints similar to the `deleteOne` and `deleteMany` functions.
